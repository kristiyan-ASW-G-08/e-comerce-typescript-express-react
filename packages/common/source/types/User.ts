export default interface User {
  email: string;
  date: string;
}
