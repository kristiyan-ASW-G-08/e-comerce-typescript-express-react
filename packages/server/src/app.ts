import express, { Request, Response, NextFunction, Application } from 'express';

import bodyParser from 'body-parser';
import helmet from 'helmet';
import compression from 'compression';
import userRoutes from '@users/routes';
import errorHandler from '@customMiddleware/errorHandler';
const app: Application = express();

app.use(helmet());
app.use(compression());
app.use(bodyParser.json());
app.use(
  bodyParser.urlencoded({
    extended: true,
  }),
);
app.use((req: Request, res: Response, next: NextFunction): void => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader(
    'Access-Control-Allow-Methods',
    'OPTIONS, GET, POST, PUT, PATCH, DELETE',
  );
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
  next();
});

app.use('/images', express.static('./images'));

app.use(userRoutes);

//@ts-ignore

app.use(errorHandler);

export default app;
