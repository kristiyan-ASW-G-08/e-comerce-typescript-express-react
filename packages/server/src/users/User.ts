import mongoose, { Schema } from 'mongoose';
import UserType from '@customTypes/UserType';
import duplicationErrorHandler from '@customMiddleware/duplicationErrorHandler';
import uniqueValidator from 'mongoose-unique-validator';

const UserSchema: Schema<UserType> = new Schema<UserType>({
  email: { required: true, type: String, minlength: 3, unique: true },
  password: { required: true, type: String, minlength: 12 },
  isConfirmed: { type: Boolean, default: true },
  date: {
    type: Date,
    default: Date.now,
  },
});

//@ts-ignore
UserSchema.plugin(uniqueValidator);

export default mongoose.model<UserType>('User', UserSchema);
