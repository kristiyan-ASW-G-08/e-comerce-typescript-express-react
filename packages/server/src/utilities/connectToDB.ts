import mongoose from 'mongoose';
import logger from '@utilities/logger';

const connectToDB = async (mongoURI: string): Promise<void> => {
  try {
    await mongoose.connect(mongoURI, {
      //@ts-ignore
      useNewUrlParser: true,
    });
  } catch (err) {
    logger.error(`MongoDB Connection Error: ${err}`);
  }
};
export default connectToDB;
