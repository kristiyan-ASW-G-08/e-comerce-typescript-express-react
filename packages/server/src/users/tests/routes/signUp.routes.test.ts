import mongoose from 'mongoose';
import request from 'supertest';
import bcrypt from 'bcryptjs';
import app from '@src/app';
import User from '@users/User';
import connectToDB from '@utilities/connectToDB';
import sendEmail from '@utilities/sendEmail';

const port = process.env.PORT || 8080;

describe('userRoutes', () => {
  const { MONGO_USER, MONGO_PASSWORD, MONGO_DATABASE } = process.env;
  const mongoURI = `mongodb+srv://${MONGO_USER}:${MONGO_PASSWORD}@cluster0.ol9wi.mongodb.net/${MONGO_DATABASE}?retryWrites=true`;
  beforeAll(async () => {
    await mongoose.disconnect();
    await connectToDB(mongoURI);
    app.listen(port);
    await User.deleteMany({}).exec();
  });
  afterEach(async () => {
    await User.deleteMany({}).exec();
  });
  afterAll(async () => {
    await mongoose.disconnect();
  });
  const email = 'testmail@mail.com';
  const password = 'testPassword';
  const invalidEmail = 'testmail';
  const invalidPassword = '1234';
  describe('/users', () => {
    it('should create a new user', async () => {
      expect.assertions(1);
      const response = await request(app)
        .post('/users')
        .send({
          email,
          password,
          confirmPassword: password,
        });
      expect(response.status).toBe(201);
    });
  });
  it("should throw an error with a status of 400: BadRequest when the req body doesn't pass validation", async (): Promise<
    void
  > => {
    expect.assertions(2);
    const response = await request(app)
      .post('/users')
      .send({
        email: invalidEmail,
        password: invalidPassword,
        confirmPassword: password,
      });
    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
    // expect(sendEmail).not.toHaveBeenCalled();
  });
  it("should throw an error with a status of 400: BadRequest when the req body doesn't pass validation", async (): Promise<
    void
  > => {
    expect.assertions(2);
    const response = await request(app).post('/users');
    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
    // expect(sendEmail).not.toHaveBeenCalled();
  });
  it('should throw an error with a status of 409: Conflict when the user credentials are already taken', async (): Promise<
    void
  > => {
    expect.assertions(1);
    await User.insertMany({ email, password });
    const response = await request(app)
      .post('/users')
      .send({
        email,
        password,
        confirmPassword: password,
      });
    expect(response.status).toBe(409);
    // expect(sendEmail).not.toHaveBeenCalled();
  });
});
