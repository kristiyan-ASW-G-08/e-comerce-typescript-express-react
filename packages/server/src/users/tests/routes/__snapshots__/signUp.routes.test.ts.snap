// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userRoutes should throw an error with a status of 400: BadRequest when the req body doesn't pass validation 1`] = `
Object {
  "data": Object {
    "validationErrors": Array [
      Object {
        "message": "email must be a valid email",
        "path": "email",
      },
      Object {
        "message": "password must be at least 12 characters",
        "path": "password",
      },
      Object {
        "message": "Passwords don't match",
        "path": "confirmPassword",
      },
    ],
  },
}
`;

exports[`userRoutes should throw an error with a status of 400: BadRequest when the req body doesn't pass validation 2`] = `
Object {
  "data": Object {
    "validationErrors": Array [
      Object {
        "message": "email is a required field",
        "path": "email",
      },
      Object {
        "message": "password is a required field",
        "path": "password",
      },
      Object {
        "message": "confirmPassword is a required field",
        "path": "confirmPassword",
      },
    ],
  },
}
`;
